examples/amm/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+    type Message = Message;
+    type SessionCall = ();
+    type Response = ();
+    type SessionState = ();
+}
+
+impl ServiceAbi for AmmAbi {
+    type Query = Request;
+    type QueryResponse = Response;
+    type Parameters = Parameters;
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum OperationType {
+    Swap {
+        owner: AccountOwner,
+        input_token_idx: u32,
+        input_amount: Amount,
+        output_token_idx: u32,
+    },
+    AddLiquidity {
+        owner: AccountOwner,
+        token0_amount: Amount,
+        token1_amount: Amount,
+    },
+    RemoveLiquidity {
+        owner: AccountOwner,
+        input_token_idx: u32,
+        input_amount: Amount,
+        other_token_idx: u32,
+    },
+}
+
+scalar!(OperationType);
+
+/// Operations that can be sent to the application.
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Operation {
+    ExecuteOperation { operation: OperationType },
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub enum Message {
+    ExecuteOperation { operation: OperationType },
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub enum ApplicationCall {
+    ExecuteOperation { operation: OperationType },
+}
+
+pub fn calculate_output_amount(
+    input_amount: Amount,
+    input_pool_balance: Amount,
+    output_pool_balance: Amount,
+) -> Result<Amount, AmmError> {
+    if input_pool_balance == Amount::ZERO || output_pool_balance == Amount::ZERO {
+        return Err(AmmError::InvalidPoolBalanceError);
+    }
+
+    let one_amount_bigint = BigUint::from_u128(<Amount as Into<u128>>::into(Amount::ONE))
+        .expect("Couldn't generate Amount::ONE in bigint");
+    let input_amount_bigint = BigUint::from_u128(<Amount as Into<u128>>::into(input_amount))
+        .expect("Couldn't generate input_amount in bigint");
+    let output_pool_balance_bigint =
+        BigUint::from_u128(<Amount as Into<u128>>::into(output_pool_balance))
+            .expect("Couldn't generate output_pool_balance in bigint");
+    let input_pool_balance_bigint =
+        BigUint::from_u128(<Amount as Into<u128>>::into(input_pool_balance))
+            .expect("Couldn't generate input_pool_balance in bigint");
+
+    // To always maintain the precision, whenever we multiply we divide by Amount::ONE
+    let numerator_bigint =
+        (input_amount_bigint.clone() * output_pool_balance_bigint) / one_amount_bigint.clone();
+    let denominator_bigint = input_pool_balance_bigint + input_amount_bigint;
+
+    if denominator_bigint.is_zero() {
+        return Err(AmmError::DivisionByZero);
+    }
+
+    // To always maintain the precision, whenever we divide, we always multiply
+    // by Amount::ONE before dividing
+    let output_amount_bigint = (numerator_bigint * one_amount_bigint) / denominator_bigint;
+    let output_amount = Amount::from(
+        output_amount_bigint
+            .to_u128()
+            .expect("Couldn't convert output_amount_bigint to u128"),
+    );
```

Comment:
Just for myself, because I can't do that in my head:

We want `(in_pool + in_amt) * (out_pool - out_amt) == in_pool * out_pool`,
i.e. `out_pool - out_amt == (in_pool * out_pool) / (in_pool + in_amt)`.
So `out_amt == out_pool - (in_pool * out_pool) / (in_pool + in_amt)`.
That's why `out_amt == out_pool * in_amt / (in_pool + in_amt)`.

Would be good to have that as a comment somewhere, with an explanation. (With the correct identifiers maybe. Or more mathematical notation?)

Reply:
Sounds good! I have a couple papers with that math somewhere too lol
I'll add a comment with it